### DISCUSSION TITLE
Selective actor, critic freezing while sharing the same context in loss function

### DISCUSSION THREAD

**Discussion by tianenchong** [ml-explore/mlx#2353] [Q&A] [answered]
I was wondering if this is doable in mlx: joint loss update while freezing a particular part of the layer (either actor or critic). Freeze here means freezing the gradient update for that layer while allowing the gradient to flow back to transformer model of shared embedded context. if stop gradient is used, it will then block the flow of gradient back to shared embedded context, which is undesirable.

```
       shared embedded context
                  /     \
    actor (freeze)     actor
               |         |
            critic    critic (freeze)

joint_loss = actor_loss + critic_loss        
```

If setting one of the layer as non-trainable parameters in the optimizer, the worse case is the loss function get called twice, once for each side of the branch, which mean we have to discard the idea of join loss. The compute resources then become 2x, as shared embedded resources get called twice in each loss.

**Comment by angeloskath** [comment, marked as answer]
This is totally doable in MLX. You have to make the forward a tiny bit more complicated in order to not use the utility `nn.value_and_grad` but it should still be less than 30 lines of code. Let me try to sketch it below.

```python
embedder = MyEmbedder(...)
actor = MyActor(...)
critic = MyCritic(...)

def loss(embedder_params, actor_params, critic_params, input, targets):
    embedder.update(embedder_params)
    actor.update(actor_params)
    critic.update(critic_params)
    x = embdedder(input)
    y = actor(x)
    z = critic(y)
    return loss_functions(x, y, z, targets)

opt_embedder = optimizers.Adam(...)
opt_actor = optimizers.Adam(...)
opt_critic = optimizers.Adam(...)

state = [embedder, actor, critic, opt_embedder.state, opt_actor.state, opt_critic.state]

@partial(mx.compile, inputs=state, outputs=state)
def step_actor(inputs, targets):
    loss, (embedder_grads, actor_grads) = mx.value_and_grad(loss, argnums=(0, 1))(
        embedder.trainable_parameters(),
        actor.trainable_parameters(),
        critic.trainable_parameters(),
        inputs,
        targets,
    )
    opt_embedder.update(embedder, embedder_grads)
    opt_actor.update(actor, actor_grads)
    return loss

@partial(mx.compile, inputs=state, outputs=state)
def step_critic(inputs, targets):
    loss, (embedder_grads, critic_grads) = mx.value_and_grad(loss, argnums=(0, 2))(
        embedder.trainable_parameters(),
        actor.trainable_parameters(),
        critic.trainable_parameters(),
        inputs,
        targets,
    )
    opt_embedder.update(embedder, embedder_grads)
    opt_critic.update(critic, critic_grads)
    return loss
```

Obviously I would remove compile to check everything works and then put it back but I think this should get you close if not work outright.

**Reply by tianenchong** [reply]
thank you so much!

**Reply by angeloskath** [reply]
By the way another way to write the above would be to use the lazy evaluation of MLX so you can get all the gradients but then only use the ones you want. Since they will be dropped they won't be evaluated. You can probably save a few lines from the above.

A way to filter the gradients would be 

```python
model = MyModel(...) # has model.embedder, model.actor, etc

def step_actor(inputs, targets)
    loss, grads = nn.value_and_grad(model, loss_fn)(inputs, targets)
    opt_embedder.update(model.embedder, grads["embedder"])
    opt_actor.update(model.actor, grads["actor"])
    return loss
```

**Reply by tianenchong** [reply]
I was wondering since we are dealing with worst case here with 2 separate step loss, in terms of step_actor and step_critic. How about fusing them in 1 single pass by replacing the frozen model with target model. this target model gets updated regularly. So, while back propagating, they will still be reasonably stable.

```
       shared embedded context
                  /     \
       target actor    actor
               |         |
            critic    target critic

joint_loss = actor_loss + critic_loss        
```
